### PR TITLE
Doc updates for 3.0 release

### PR DESCRIPTION
**Pull request by jhamman**
This PR includes a fairly large reorganization of the Zarr-Python documentation in preparation for the 3.0 release. Major changes include:

- new top level `quickstart` page
- break up `tutorial.rst`. See `user-guide/{parts}.rst`
  - use ipython directives to run examples at doc build time
  - new user guide sections for advanced topics (async, extensions, etc.)
- add new site section for developers (contributors guide, roadmap, etc.)

#### TODOs
- [x] Get feedback and decide on what to do with the docs in `todo.rst`. These are mostly features that have not been ported to v3.
- [x] I would like to migrate the content from #2102 into `user-guide/whatsnew_v3.rst`.

Closes: #1765, #1767, #1770, #1798, #2550

### CODE REVIEW

File: docs/user-guide/todo.rst [hunk 1]
```diff
@@ -0,0 +1,566 @@
+TODO
```
Comment (jhamman) [inline docs/user-guide/todo.rst:1]: this page is where all the problem children are!

File: .readthedocs.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@ build:
 
 sphinx:
   configuration: docs/conf.py
-  fail_on_warning: true
+  fail_on_warning: false
```
Comment (jhamman) [inline .readthedocs.yaml:10]: todo: revert this change before merging!

File: docs/user-guide/extending.rst [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+Extending Zarr
+==============
+
+Zarr-Python 3 was designed to be extensible. This means that you can extend
+the library by writing custom classes and plugins. Currently, Zarr can be extended
+in the following ways:
+
+1. Writing custom stores
+2. Writing custom codecs
```
Comment (jhamman) [inline docs/user-guide/extending.rst:10]: @normanrz - could you suggest a short bit of text on how the codec system is meant to work.

File: docs/index.rst [hunk 4]
```diff
@@ -94,13 +100,17 @@ Zarr is a file storage format for chunked, compressed, N-dimensional arrays base
         Contributor's Guide
```
Comment (joshmoore) [inline docs/index.rst:100]: Could see this point to the Developer's Guide to parallel the top-row.

**Comment by jhamman** [comment]
> could this PR be split up into smaller bits where possible?

This is possible but I think the size here is appropriate given the intent and scope of the work. I'm fine with #2570 but before I split things out any further, I'd like to get feedback on the structure of the new docs. I wrote very few new lines here, it is mostly a reshuffle of existing content. 


----

To clarify, at this stage, I'm looking for "30%" feedback. What do you think of the new structure? Do you think this will make the docs easier to navigate, maintain, and/or extend? What other sections should we add (in follow up PRs)?

**Comment by dstansby** [comment]
Had a look, and I'm 👍 👍 for splitting the tutorial up into sub-pages. I think I'm 👍 on renaming it to "guide" too. So I think at a high level this looks great!

Sorry for my negativity further up - I got a bit spooked by some of the content changes and didn't look past them to the tutorial rearrangement.

File: docs/index.rst (hunk 4 above)
Comment (jhamman) [reply docs/index.rst:100]: @joshmoore - I'm having trouble parsing your comment here. Can you restate it?

File: docs/user-guide/storage.rst [hunk 5]
```diff
@@ -40,55 +49,52 @@ Local Store
...
+      storage_options={"anon": True}
+   )
+   zarr.open_group(store=store, mode='r')
 
 Memory Store
 ~~~~~~~~~~~~
 
-The :class:`zarr.storage.RemoteStore` a in-memory store that allows for serialization of
+The :class:`zarr.storage.FsspecStore` a in-memory store that allows for serialization of
 Zarr data (metadata and chunks) to a dictionary.
```
Comment (maxrjones) [inline docs/user-guide/storage.rst:91]:
```suggestion
The :class:`zarr.storage.MemoryStore` allows for serialization of Zarr data (metadata and chunks) to an im-memory dictionary.
```

File: docs/user-guide/storage.rst [hunk 6]
```diff
@@ -40,55 +49,52 @@ Local Store
...
-   >>> zarr.open(store=store)
-   <Group file://data/foo/bar>
+   store = zarr.storage.LocalStore("data/foo/bar", read_only=True)
+   zarr.open(store=store, mode='r')
 
 Zip Store
 ~~~~~~~~~
 
 The :class:`zarr.storage.ZipStore` stores the contents of a Zarr hierarchy in a single
 Zip file. The `Zip Store specification_` is currently in draft form.
```
Comment (maxrjones) [inline docs/user-guide/storage.rst:61]: This needs a link in the form of `Zip Store specification <insert link>`_

File: docs/user-guide/storage.rst [hunk 7]
```diff
@@ -40,55 +49,52 @@ Local Store
...
 Remote Store
 ~~~~~~~~~~~~
 
-The :class:`zarr.storage.RemoteStore` stores the contents of a Zarr hierarchy in following the same
+The :class:`zarr.storage.FsspecStore` stores the contents of a Zarr hierarchy in following the same
 logical layout as the ``LocalStore``, except the store is assumed to be on a remote storage system
 such as cloud object storage (e.g. AWS S3, Google Cloud Storage, Azure Blob Store). The
-:class:`zarr.storage.RemoteStore` is backed by `Fsspec_` and can support any Fsspec backend
-that implements the `AbstractFileSystem` API,
+:class:`zarr.storage.FsspecStore` is backed by `Fsspec_` and can support any Fsspec backend
```
Comment (maxrjones) [inline docs/user-guide/storage.rst:74]:
```suggestion
:class:`zarr.storage.FsspecStore` is backed by `Fsspec <https://filesystem-spec.readthedocs.io/en/latest/>`_ and can support any Fsspec backend
```

File: docs/user-guide/v3_migration.rst [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+Migration Guide
+---------------
+
+The following sections provide details on the most important changes in Zarr-Python 3.
+
+The Array class
+~~~~~~~~~~~~~~~
+
+1. Disallow direct construction - use :func:`zarr.open_array` or :func:`zarr.create_array`
```
Comment (maxrjones) [inline docs/user-guide/v3_migration.rst:93]:
```suggestion
1. Disallow direct construction - use :func:`zarr.open_array` or :func:`zarr.Group.create_array`
```

**Review by maxrjones — COMMENTED** [review]
This is wonderful @jhamman! I really like the new organization and finally understand vindex vs. oindex 😅

I left a few in-line suggestions and have some general suggestions below, but this would be good to go after a bit of cleanup even if you don't take any of them.

 - I'm not sure the comparisons to h5py add a lot of value. I would put them as Notes after rather than in the main part of the explanation, or just remove them to not assume familiarity with h5py.
- I think using more subsections in groups would be helpful, similar to arrays (i.e., Creating groups, Persistent Groups)
- I think Attributes would be better off as a separate section rather than nested under "working with groups" since it pertains to groups and arrays
- Since storage is implicit in the arrays and groups sections, I recommend renaming "storage section" to "working with storage backends" or "working with storage"
- I assume the TODOs in array compression are known issues?
- I thought extensibility was a major reason for Zarr V3? I think it's worth being as comprehensive as possible in the motivations list of the migration guide since the major version bump will place work on users.
- In the v3 migration guide, it would be extra nice if you could include a pointer to those who rely on the deprecated stores on what to do. Perhaps a link to the extension page?
- It's worth being super clear up-front that the migration page isn't comprehensive. E.g., aren't there some small but meaningful changes like listdir() -> list_dir?
- There are some existing references to `RemoteStore` that are beyond where GH supports in-line suggestions

File: docs/user-guide/v3_migration.rst [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+If you need to use the latest Zarr-Python 2 release, you can install it with:
+
+.. code-block:: console
+
+    $ pip install "zarr==2.*"
+
+Migration Guide
+---------------
+
+The following sections provide details on the most important changes in Zarr-Python 3.
```
Comment (normanrz) [inline docs/user-guide/v3_migration.rst:88]: We should add that `zarr_format=3` is the new default.

File: docs/user-guide/arrays.rst [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+Persistent arrays
+-----------------
+
+In the examples above, compressed data for each chunk of the array was stored in
+main memory. Zarr arrays can also be stored on a file system, enabling
+persistence of data between sessions. For example:
+
+.. ipython:: python
+
+   z1 = zarr.open('data/example.zarr', mode='w', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:67]: Should we replace all uses of `zarr.open`, `zarr.array`, `zarr.group` in favor of `zarr.{create,open}_{array,group}` once #2463 lands?

File: docs/user-guide/arrays.rst [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+   a = np.arange(10000000, dtype='i4').reshape(10000, 1000)
+   z = zarr.array(a, chunks=(1000, 100))
+   z.shape
+   z.append(a)
+   z.append(np.vstack([a, a]), axis=1)
+   z.shape
+
+.. _tutorial_compress:
+
+Compressors
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:145]: I think we should add a section for sharding.

File: docs/user-guide/v3_migration.rst (hunk 9 above)
Comment (normanrz) [reply docs/user-guide/v3_migration.rst:88]: Another thing that is missing, is the new `zarr.config` mechanism

Comment (normanrz) [reply docs/user-guide/v3_migration.rst:88]: I added some text for these items. Feel free to edit liberally.

File: docs/user-guide/arrays.rst
Comment (normanrz) [inline docs/user-guide/arrays.rst:1]: Where is sharding? 😆

File: docs/index.rst (hunk 4 above)
Comment (joshmoore) [reply docs/index.rst:100]: The box at the bottom is labeled and links to "Contributor's Guide" while the link at the top is labeled and links to "Developer's Guide". It'd err on the side of using Dev Guide for both.

File: src/zarr/core/config.py
Comment (normanrz) [inline src/zarr/core/config.py:1]: Seems like someone has already added some docs for the config in `main` in the meantime.

File: docs/quickstart.rst [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+.. ipython:: python
+
+    # Store the array in a directory on disk
+    z = zarr.open(
+        'data/example-1.zarr',
+        mode='w', shape=(100, 100), chunks=(10, 10), dtype='f4'
+    )
+    z[:, :] = np.random.random((100, 100))
+
+    print("Array stored at 'data/example-1.zarr'")
```
Comment (dcherian) [inline docs/quickstart.rst:66]:
```suggestion
    # write to the array
    z[:, :] = np.random.random((100, 100))
```

File: docs/user-guide/arrays.rst [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+Creating an array
+-----------------
+
+Zarr has several functions for creating arrays. For example:
+
+.. ipython:: python
+
+   import zarr
+
+   z = zarr.zeros((10000, 10000), chunks=(1000, 1000), dtype='i4')
```
Comment (dcherian) [inline docs/user-guide/arrays.rst:14]:
```suggestion
   z = zarr.open((10000, 10000), chunks=(1000, 1000), dtype='i4')
```

I don't think the first example on this page should initialize the data at creation. The next section talks about writing values, so simply `open` or whatever.

File: docs/user-guide/arrays.rst (hunk 11 above)
Comment (dcherian) [reply docs/user-guide/arrays.rst:145]: top-level note on the differences between V2 and V3 would be good. (I'm sure this is on your todo list, just a reminder)

File: docs/user-guide/consolidated_metadata.rst
Comment (dcherian) [inline docs/user-guide/consolidated_metadata.rst:1]: The link to the spec in the first line is broken.

File: docs/user-guide/arrays.rst (hunk 13 above)
Comment (dcherian) [reply docs/user-guide/arrays.rst:14]: Also I think the "persistent storage" bit should be moved up here. Most new users are looking for a way to store data on disk, starting off with MemoryStore is confusing (from my personal experience)

Comment (normanrz) [reply docs/user-guide/arrays.rst:14]: I agree. The main point of using zarr over numpy is persistence

File: docs/user-guide/arrays.rst
Comment (jhamman) [reply docs/user-guide/arrays.rst:1]: I've added a Sharding section with a "coming soon" label.

**Comment by dstansby** [comment]
What's the plan with this PR - is it going to be split into several smaller PRs to make it easier to review?

**Comment by jhamman** [comment]
@dstansby - I don't plan on splitting this up into multiple PRs. I'd like to merge this in more or less its current state then wrap up the TODOs and placeholder sections in smaller PRs to come after (I've already started a new branch integrating #2463 on top of this). Are there parts that you would like to review before we merge?

**Comment by dstansby** [comment]
Yes, ideally I'd like to review all of it 😄 I know it's a little bit more effort to split it up, but I think it will make it much easier to get different parts reviewed and catch any mistakes and make improvements. At the moment, reviewing a +/- 2000 line diff is going to take ages. I'm happy to put in the effort to split it up if you don't mind?

**Comment by normanrz** [comment]
This PR largely rewrites the user guide. For a review, I would just read the new user guide from top to bottom instead of looking at the diff. That should cover most of this PR.

**Comment by dstansby** [comment]
I disagree - changes in this PR that could be independent PRs are:

- Moving info to `docs/about.rst`
- Using IPython for syntax highlighting
- Changes/updates to the contributing guide
- Moving pages to docs/developers
- Adding a new quickstart page
- Replacing the tutorial with new user guides
- Adding a new "extending zarr" page
- Updating the installation page
- Adding a new "performance" page
- Updating the "storage" user guide
- Adding a new v3 migration guide
- Adding a new v3 todos guide

By putting these altogether in one big PR it makes them much harder to review because the reviewer has to keep the context of all the changes in their head at once. If instead split up into multiple PRs:

- the time taken to review each one should be much shorter than the total time needed to (properly) review this PR
- it will be much easier to catch mistakes and make improvements on smaller PRs
- it's much more likely multiple core devs will have time to review differnet smaller PRs, which goes with the spirit of [Merge Only Changes You Understand](https://github.com/zarr-developers/governance/blob/main/CORE_DEV_GUIDE.md#merge-only-changes-you-understand) ("Code doesn't merely have to work, but should be understood by multiple core developers.")

I very strongly believe this should be split up into multiple PRs, instead of being a large collection of multiple concurrent changes to the docs. Again, I'm happy to put in the effort to split this into multiple PRs and do the required rebasing.

**Comment by normanrz** [comment]
While I usually also favor smaller PRs, I don't think that is the case for this docs PR. Splitting up the PR and reviewing in isolation won't necessarily result in a more cohesive documentation.  Again, I think for an overhaul this large, I think it makes more sense to look at the new docs and review whether it is easy-to-understand, comprehensive and correct than to examine the diff. Essentially, you could ignore that old docs existed. 

I think it should be @jhamman's call whether he wants to split up the PR.

File: docs/quickstart.rst [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    from numcodecs import Blosc
+
+    z = zarr.open(
+        "data/example-3.zarr",
+        mode="w", shape=(100, 100),
+        chunks=(10, 10), dtype="f4",
+        compressor=Blosc(cname="zstd", clevel=3, shuffle=Blosc.SHUFFLE),
+        zarr_format=2
+    )
+    z[:, :] = np.random.random((100, 100))
```
Comment (normanrz) [inline docs/quickstart.rst:77]:
```suggestion
    z = zarr.create_array(
        "data/example-3.zarr",
        mode="w", shape=(100, 100),
        chunks=(10, 10), dtype="f4",
        compressors=(zarr.codecs.BloscCodec(),),
    )
    z[:, :] = np.random.random((100, 100))
```

File: docs/quickstart.rst [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+Zarr supports persistent storage to disk or cloud-compatible backends. While examples above
+utilized a :class:`zarr.storage.LocalStore`, a number of other storage options are available,
+including the :class:`zarr.storage.ZipStore` and :class:`zarr.storage.FsspecStore`.
+
+.. ipython:: python
+
+    # Store the array in a ZIP file
+    store = zarr.storage.ZipStore("data/example-3.zip", mode='w')
+
+    z = zarr.open(
```
Comment (normanrz) [inline docs/quickstart.rst:119]:
```suggestion
    z = zarr.create_array(
```

File: docs/quickstart.rst [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+    store.close()
+
+To open an existing array:
+
+.. ipython:: python
+
+    # Open the ZipStore in read-only mode
+    store = zarr.storage.ZipStore("data/example-3.zip", read_only=True)
+
+    z = zarr.open(store, mode='r')
```
Comment (normanrz) [inline docs/quickstart.rst:140]:
```suggestion
    z = zarr.open_array(store, mode='r')
```

File: docs/quickstart.rst [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+`gcsfs <https://gcsfs.readthedocs.io>`_.
+
+For example, to use S3:
+
+.. ipython:: python
+   :verbatim:
+
+    import s3fs
+
+    z = zarr.open("s3://example-bucket/foo", mode="w", shape=(100, 100), chunks=(10, 10))
```
Comment (normanrz) [inline docs/quickstart.rst:159]:
```suggestion
    z = zarr.create_array("s3://example-bucket/foo", mode="w", shape=(100, 100), chunks=(10, 10))
```

File: docs/user-guide/arrays.rst [hunk 18]
```diff
@@ -0,0 +1,528 @@
...
+
+Resizing and appending
+----------------------
+
+A Zarr array can be resized, which means that any of its dimensions can be
+increased or decreased in length. For example:
+
+.. ipython:: python
+
+   z = zarr.zeros(
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:138]:
```suggestion
   z = zarr.create_array(
```

File: docs/user-guide/arrays.rst [hunk 19]
```diff
@@ -0,0 +1,528 @@
...
+any way. If one or more dimensions are shrunk, any chunks falling outside the
+new array shape will be deleted from the underlying store.
+
+:func:`zarr.Array.append` is provided as a convenience function, which can be
+used to append data to any axis. E.g.:
+
+.. ipython:: python
+
+   a = np.arange(10000000, dtype='i4').reshape(10000, 1000)
+   z = zarr.array(store="data/example-5", data=a, chunks=(1000, 100))
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:158]:
```suggestion
   z = zarr.create_array(store="data/example-5", data=a, chunks=(1000, 100))
```

File: docs/user-guide/arrays.rst [hunk 20]
```diff
@@ -0,0 +1,528 @@
...
+argument accepted by all array creation functions. For example:
+
+.. ipython:: python
+
+   from numcodecs import Blosc
+
+   compressor = None  # TODO: Blosc(cname='zstd', clevel=3, shuffle=Blosc.BITSHUFFLE)
+   data = np.arange(100000000, dtype='i4').reshape(10000, 10000)
+   # TODO: remove zarr_format
+   z = zarr.array(store="data/example-6.zarr", data=data, chunks=(1000, 1000), compressor=compressor, zarr_format=2)
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:182]:
```suggestion
   compressor = None  # TODO: zarr.codecs.BloscCodec(cname='zstd', clevel=3, shuffle=Blosc.BITSHUFFLE)
   data = np.arange(100000000, dtype='i4').reshape(10000, 10000)
   # TODO: remove zarr_format
   z = zarr.create_array(store="data/example-6.zarr", data=data, chunks=(1000, 1000), compressor=compressor)
```

File: docs/user-guide/arrays.rst [hunk 21]
```diff
@@ -0,0 +1,528 @@
...
+highly optimized implementations of byte- and bit-shuffle filters, which can
+improve compression ratios for some data. A list of the internal compression
+libraries available within Blosc can be obtained via:
+
+.. ipython:: python
+
+   from numcodecs import blosc
+
+   blosc.list_compressors()
+
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:218]: The default now is Zstd. Also, I think we should link out for the details of the blosc codec.

File: docs/user-guide/arrays.rst [hunk 22]
```diff
@@ -0,0 +1,528 @@
...
+   from numcodecs import Zstd
+
+   z = zarr.array(
+       store="data/example-7.zarr",
+       data=np.arange(100000000, dtype='i4').reshape(10000, 10000),
+       chunks=(1000, 1000),
+       compressor=Zstd(level=1),
+       zarr_format=2  # TODO: remove zarr_format
+   )
```
Comment (normanrz) [inline docs/user-guide/arrays.rst:232]:
```suggestion
   z = zarr.create_array(
       store="data/example-7.zarr",
       data=np.arange(100000000, dtype='i4').reshape(10000, 10000),
       chunks=(1000, 1000),
       compressor=zarr.codecs.ZstdCodec(level=1)
   )
```

File: docs/user-guide/async.rst [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+The usage of the async API mirrors the synchronous API.
+
+.. ipython:: python
+
+   import asyncio
+   import numpy as np
+   import zarr.api.asynchronous as async_zarr
+
+   # create a new group using the asynchronous API
+   root = await async_zarr.group(attributes={'foo': 'bar'})
```
Comment (normanrz) [inline docs/user-guide/async.rst:16]:
```suggestion
   root = await async_zarr.create_group(attributes={'foo': 'bar'})
```

File: docs/user-guide/attributes.rst [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+Zarr arrays and groups support custom key/value attributes, which can be useful for
+storing application-specific metadata. For example:
+
+.. ipython:: python
+
+   root = zarr.group()
+
+   root.attrs['foo'] = 'bar'
+
+   z = root.zeros(name='zzz', shape=(10000, 10000))
```
Comment (normanrz) [inline docs/user-guide/attributes.rst:15]:
```suggestion
   z = root.create_array(name='zzz', shape=(10000, 10000), dtype="i8")
```

File: docs/user-guide/groups.rst [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+groups can be stored in memory, on disk, or via other storage systems that
+support a similar interface.
+
+To create a group, use the :func:`zarr.group` function:
+
+.. ipython:: python
+
+   import zarr
+
+   root = zarr.group()
```
Comment (normanrz) [inline docs/user-guide/groups.rst:20]:
```suggestion
   root = zarr.create_group()
```

File: docs/user-guide/groups.rst [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+.. ipython:: python
+
+   foo = root.create_group('foo')
+   bar = foo.create_group('bar')
+
+Groups can also contain arrays, e.g.:
+
+.. ipython:: python
+
+   z1 = bar.zeros(name='baz', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
```
Comment (normanrz) [inline docs/user-guide/groups.rst:35]:
```suggestion
   z1 = bar.create_array(name='baz', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
```

File: docs/user-guide/groups.rst [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+   :suppress:
+
+   rm -r data/group.zarr
+
+.. ipython:: python
+
+   root = zarr.open_group('data/group.zarr', mode='w')
+   root
+
+   z = root.zeros(name='foo/bar/baz', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
```
Comment (normanrz) [inline docs/user-guide/groups.rst:73]:
```suggestion
   z = root.create_array(name='foo/bar/baz', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
```

File: docs/user-guide/groups.rst [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+
+Array and group diagnostics
+---------------------------
+
+Diagnostic information about arrays and groups is available via the ``info``
+property. E.g.:
+
+.. ipython:: python
+
+   root = zarr.group()
```
Comment (normanrz) [inline docs/user-guide/groups.rst:92]:
```suggestion
   root = zarr.create_group()
```

File: docs/user-guide/performance.rst [hunk 29]
```diff
@@ -0,0 +1,332 @@
...
+for a 2-dimensional array, if you only ever take slices along the first
+dimension, then chunk across the second dimension. If you know you want to chunk
+across an entire dimension you can use ``None`` or ``-1`` within the ``chunks``
+argument, e.g.:
+
+.. ipython:: python
+
+   import zarr
+
+   z1 = zarr.zeros((10000, 10000), chunks=(100, None), dtype='i4')
```
Comment (normanrz) [inline docs/user-guide/performance.rst:33]:
```suggestion
   z1 = zarr.create_array((10000, 10000), chunks=(100, None), dtype='i4')
```
A bunch of these in this file.

File: docs/user-guide/performance.rst [hunk 30]
```diff
@@ -0,0 +1,332 @@
...
+filters (e.g., byte-shuffle) have been applied.
+
+.. _tutorial_chunks_empty_chunks:
+
+Empty chunks
+~~~~~~~~~~~~
+
+It is possible to configure how Zarr handles the storage of chunks that are "empty"
+(i.e., every element in the chunk is equal to the array's fill value). When creating
+an array with ``write_empty_chunks=False``, Zarr will check whether a chunk is empty before compression and storage. If a chunk is empty,
```
Comment (normanrz) [inline docs/user-guide/performance.rst:102]:
```suggestion
an array with the config ``array.write_empty_chunks=False``, Zarr will check whether a chunk is empty before compression and storage. If a chunk is empty,
```

File: docs/user-guide/v3_todos.rst [hunk 31]
```diff
@@ -0,0 +1,222 @@
+3.0 TO DOs
+==========
+
+.. warning::
+   As noted in the `3.0 Migration Guide <user-guide/v3_migration>`_, there are still a few
+   features that were present in Zarr-Python 2 that are not yet ported to Zarr-Python 3.
```
Comment (normanrz) [inline docs/user-guide/v3_todos.rst:6]: Maybe we should state that some features might not be ported at all?

**Comment by dstansby** [comment]
> I think for an overhaul this large, I think it makes more sense to look at the new docs and review whether it is easy-to-understand, comprehensive and correct than to examine the diff. Essentially, you could ignore that old docs existed.

I would agree if this PR was *only* an entirely new set of docs, but it isn't, it's making a bunch of changes and also deleting a bunch of old docs. So we need to make sure that nothing is being lost, and that the changes are sensible.

Perhaps a good halfway house would be to move some of the changes into separate PRs to make it easier to see what's changed, and then have a large (but still smaller than this) PR that contains only deleting the old tutorial, and all the new guides that replace them?

Concretely, changes that I think would benefit from being their own PR:

- Moving info to docs/about.rst (already done in https://github.com/zarr-developers/zarr-python/pull/2566)
- Moving pages to docs/developers
- Using IPython for syntax highlighting
- Changes/updates to the contributing guide
- Updating the installation page

With a few of us available I think we could review/improve/merge thsese all pretty quickly, then leaving a much cleaner and easier to review PR here.

**Comment by jhamman** [comment]
I'm going to close this PR and open up a string of smaller PRs. More soon.
